src/server/mod.rs

```diff
@@ -96,6 +106,19 @@ impl<I> Server<I, ()> {
     }
 }
 
+impl Server<tokio_tcp::Incoming, ()> {
+    /// Create a new instance from a `std::net::TcpListener` instance.
+    #[cfg(feature = "runtime")]
+    pub fn from_tcp(
+        listener: net::TcpListener,
+    ) -> Result<Builder<tokio_tcp::Incoming>, ::Error> {
```

Comment:
I thought about it some more, and I do think it makes sense to convert into the internal `AddrIncoming` type instead. The changes will be easy, and then anyone using it will be able to take advantage of the default improvements like sleeping when max FDs is reached.

The changes are to add a `pub(super) fn from_tcp()` to `AddrIncomong`. The `new` function can be split, such that after getting the `std_listener`, it just calls `AddrIncoming::from_tcp` internally. And then this function also can just call `AddrIncoming::from_tcp(listener)`.

Reply:
hey, I was about to send in a similar patch. I wrote [hyper-tls-hack](https://github.com/miquels/hyper-tls-hack) , an implementation of AddrIncoming that supports TLS using tokio-tls. It has a `from_std_listener(std_listener: StdTcpListener,....` method, which I was about to split off and add to hyper's AddrIncoming. See [the code here](https://github.com/miquels/hyper-tls-hack/blob/master/src/lib.rs#L153) .